### PR TITLE
[Platformer] Allow to automatically grab platform ledges without having to move horizontally.

### PR DESCRIPTION
**Pull request by D8H**
Without this option, players may release the Left or Right key too soon when the character is against the wall for the character to grab the platform.

This is the implementation for the (1) of this ticket:
* https://trello.com/c/zezNvn1J/269-add-an-option-to-automatically-grab-the-ledge-of-platforms

Contrary to this extension, it doesn't automatically move the character to be in range:
* https://github.com/GDevelopApp/GDevelop-extensions/pull/341

### CODE REVIEW

**Comment by 4ian** [comment]
Interesting, I've not checked the code yet, but is there any performance hit that would come from this option?

**Comment by D8H** [comment]
> Interesting, I've not checked the code yet, but is there any performance hit that would come from this option?

There should be no impact on CPU usage, but it uses 2 boolean attributes.

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 1]
```diff
@@ -69,98 +280,92 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
...
-          upperPlatform.getX() + upperPlatform.getWidth() + 2,
-          upperPlatform.getY() - 10
-        );
-        runtimeScene.renderAndStep(1000 / 60);
-
-        for (let i = 0; i < 35; ++i) {
-          object.getBehavior('auto1').simulateLeftKey();
-          runtimeScene.renderAndStep(1000 / 60);
-        }
+    it('can grab without moving and release a left ledge platform', function () {
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:283]:
```suggestion
    it('can grab without moving and release the left ledge of a platform', function () {
```

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 2]
```diff
@@ -69,98 +280,92 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
...
+      runtimeScene.renderAndStep(1000 / 60);
 
-        // Check that the object grabbed the upper platform
-        expect(object.getX()).to.be.within(
-          upperPlatform.getX() + upperPlatform.getWidth() + 0,
-          upperPlatform.getX() + upperPlatform.getWidth() + 1
-        );
-        expect(object.getY()).to.be(upperPlatform.getY());
-        expect(object.getBehavior('auto1').isGrabbingPlatform()).to.be(true);
+      // Put the character near the right ledge of the platform.
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:289]:
```suggestion
      // Put the character near the left ledge of the platform.
```

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 3]
```diff
@@ -32,7 +240,7 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
       object.setPosition(0, -100);
     });
 
-    it('can grab, and release, a platform', function () {
+    it('can grab without moving and release a right ledge platform', function () {
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:243]:
```suggestion
    it('can grab without moving and release the right ledge of a platform', function () {
```

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 4]
```diff
@@ -69,98 +280,92 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
...
-      expect(object.getX()).to.be.within(
-        platform.getX() + platform.getWidth() + 0,
-        platform.getX() + platform.getWidth() + 1
+    it('must not grab automatically a left ledge platform when facing back', function () {
+      // Put a platform.
+      const platform = addPlatformObject(runtimeScene);
+      platform.setPosition(0, -10);
+      runtimeScene.renderAndStep(1000 / 60);
+
+      // Put the character near the right ledge of the platform.
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:352]:
```suggestion
      // Put the character near the left ledge of the platform.
```

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 5]
```diff
@@ -69,98 +280,92 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
...
         );
-        expect(object.getY()).to.be(lowerPlatform.getY());
-        expect(object.getBehavior('auto1').isGrabbingPlatform()).to.be(true);
-      });
+      }
+      expect(object.getY()).to.be.above(0);
     });
 
-    it('can grab a platform and jump', function () {
+    it('must not grab automatically a right ledge platform when facing back', function () {
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:321]:
```suggestion
    it('must not grab automatically the right ledge of a platform when facing back', function () {
```

File: Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js [hunk 6]
```diff
@@ -69,98 +280,92 @@ describe('gdjs.PlatformerObjectRuntimeBehavior', function () {
...
+      expect(object.getBehavior('auto1').isFallingWithoutJumping()).to.be(true);
+      expect(object.getY()).to.be.greaterThan(platform.getY() + 5);
+    });
 
-      //Check that the object grabbed the platform
-      expect(object.getBehavior('auto1').isGrabbingPlatform()).to.be(true);
-      expect(object.getX()).to.be.within(
-        platform.getX() + platform.getWidth() + 0,
-        platform.getX() + platform.getWidth() + 1
+    it('must not grab automatically a left ledge platform when facing back', function () {
```
Comment (4ian) [inline Extensions/PlatformBehavior/tests/LadderAndCliffPlatformer.spec.js:346]:
```suggestion
    it('must not grab automatically the left ledge of a platform when facing back', function () {
```

**Review by 4ian — APPROVED** [review]
Just a few comments/titles to update! :)

**Pull request merged**